### ISSUE TITLE
Media sync between multiple beets servers

### ISSUE THREAD

**Issue by jankowiak** [beetbox/beets#271]
Implement media sync between multiple instances for beets to keep files and metadata up to date across multiple computers/servers.

**Comment by alphapapa** [comment]
I've been using MPD on multiple systems with libraries synced by Unison.  This works very well, since MPD automatically re-reads the metadata for updated files.

I wonder if a "poor man's" solution to this would be to have Beets set up to:

1.  Write all metadata back to track files immediately
2.  Automatically read metadata from modified track files (i.e. "reimport" them, I guess)

Then whenever the underlying library files change when synced by Unison (or git-annex, rsync, etc), Beets would update its database accordingly.

I'm new to Beets, so am I missing anything, or would this not work?  Thanks.

**Comment by sim590** [comment]
Hello gentlemen,

*I have noticed that this post is old, but I assume there are no solutions yet since the issue is still opened.*

Let me introduce an out-of-the box solution: use a distributed hash table to establish peer-to-peer communication between each of your devices and finally exchange list of files present on both side in order to finally sync files. This solution wouldn't need to rely on a centralized server, it would be efficient and reliable.

I'm actually planning on starting a project that would sync files in a given directory between peers using [OpenDHT][], a project I have been contributing for more than a year so far. If this solution was to be taken, it would also be a secure since beets would directly benefit from the DHT's security layer it provides. OpenDHT also has a python wrapper around it's core API.

However, I'm presently wondering if the bittorrent protocol would be sufficient and if it would suit the purpose well. I have given a look at the [bittorrent protocol][] and the c++ library implementation [libtorrent][] for which there exists a [python wrapper](https://pypi.python.org/pypi/python-libtorrent/), but it's not clear for me yet. One may ask the following questions:

- *Bittorrent's protocol is based on the usage of `.torrent` files to list the content to share. Would it be suited for dynamic changes in the list of files to share like files changing in a directory?*
- *Bittorrent uses trackers. Is the usage of such trackers mandatory for using the bittorrent protocol?*

I don't know the bittorrent protocol well for now, so I don't have answers to these questions. Also, there may have other problems I'm not aware of. However, I know how I would do it using OpenDHT. I also want to point out that the solution I consider would be designed to share files between two peers in a secure manner which would imply that both peers sharing data know each other a priori. Therefor, possible use cases:

- sharing between two devices you own;
- sharing between friends;

The way I see it, the layer implementing file sharing wouldn't *need* to be designed to enable mass file sharing. If someone has appropriate knowledge about the Bittorrent protocol, I would appreciate you to share with us (for a lack of better words :wink: ).

I think that distributed solutions should be more and more part of the developer's choices nowadays considering how critical data storage and sharing has become. Our freedom to achieve data sharing is increasingly threatened. I believe distributed and secure software is the tool to use to guarantee the users their freedom.

[OpenDHT]: http://opendht.net
[bittorrent protocol]: http://www.bittorrent.org/beps/bep_0003.html
[libtorrent]: http://www.rasterbar.com/products/libtorrent/

**Comment by sim590** [comment]
After some researching, I have come across [Syncthing][]. I think that this is really good project. Especially since they're planning on [using a DHT for device discovery][dht-syncthing]. This is indeed better suited than the bittorrent protocol because the bittorrent protocol is not designed for private file sharing between few devices, but rather permits sharing to scale to high numbers of peers. The bittorrent protocol doesn't handle authentication and such. The only downside to Syncthing is that the project is not designed to be used as library. Therefore, it's less desirable to integrate into beets.

Anyway, it's possible to use syncthing as a standalone solution. It should be recommended to beets users IMO. I also foresee the possibility of integrating it into beets, but not nicely as I said.

[Syncthing]: https://syncthing.net/
[dht-syncthing]: https://github.com/syncthing/syncthing/issues/3388

**Comment by alphapapa** [comment]
@sim590 Very thoughtful ideas, thanks.

I think we should separate the problem into two distinct issues:

1.  Discovering and connecting peers.
2.  Syncing files between peers.

DHT or Avahi solve number 1, depending on whether you want to sync over the Internet or just over a LAN.

Number 2 is the tough one.  The problem with syncing is always that of resolving conflicts.  It's likely manageable when dealing with two peers, because most of the time changes to one file will be made in one place and synced before changes are made to the same file independently; if a conflict does happen, for something like Beets, it would probably be sufficient to just take the latest version--we're just talking about metadata.  And manually resolving conflicts isn't too bad for metadata.  But if you add more than 2 peers, it starts getting much more complicated.

There's also the issue of whether we're syncing files on disk or the contents of a database.  If it's just files, then something like a thin wrapper around `unison` might be a 99% solution.  I've been using it to sync my music library for years, and I never have to think about it, it just works.  I keep metadata directly in the track files, and MPD's and Clementine's databases serve as a local cache for each system.  But if you're talking about syncing database contents, that's a whole 'nother ballgame (although a file-based patch-like series of database inserts/updates synced with Unison might even work for that).

Anyway, interesting thoughts.  I would suggest decoupling syncing from the project as much as possible: the more you can treat the data as plain files, the more that existing, proven tools can be used to do the job.

**Comment by dosoe** [comment]
Hello, I'm bursting in this topic without any technical knowledge, but is beets able to read its own database? Could it be possible to give him another beets database as an input and compare it to the 'regular' one (the one that is his according to the config file)? He would then look at the difference and then either print them out, or copy everything that is in one database and not the other (with maybe a path format as an option). Would that be technically difficult? I mean, we already have the code to read one database and we have the code to move/copy files, so we would just need to catch these chunks of code and arrange them, right? I am asking because I have a big library and a backup of this library, but I would like to even up the difference once every few weeks. Both have their beets library.

**Comment by sampsyo** [comment]
Yes, this is possible in principle! In fact, very early versions of beets had the ability to work with different libraries—back when it seemed important to treat one's iPod as a distinct "library." We never got as far as transferring information between two libraries, though.

FWIW, the hard part in all this is comparing two databases. It's surprisingly hard to do efficiently and correctly. Imagine, for example, that you change the title of a track in one copy of the database. On the next sync, how do you know whether this is a new track (and the old one was deleted) or the same track with a modification applied? Things get even trickier if you allow simultaneous changes on both copies. The rest of the system—the low-level communication between two libraries—is just implementation once we figure out those sync semantics.

**Comment by dosoe** [comment]
Well, all these recordings have ids, right? The default would be the musicbrainz ID, and I guess discogs and all the other databases we can call do have some too. With the album id and the recording id, it should be easy to identify the recordings, and I do not see why someone would change them manually (but maybe there would be a reason?). Could it be possible to use them as unique identifiers for tracks and albums (as they are meant for) for synching up and avoid the problem you just stated? This wouldn't work for tracks that have been added without being autotagged (-A), but that's a start.

**Comment by sampsyo** [comment]
Yep, MBIDs are a good disambiguator—but as you mention, it's possible to changed a track's MBID (if, for example, you import music that's not in MusicBrainz and then later it gets added), or to have multiple tracks with the same MBID.

Anyway, this sort of thing is certainly a workable idea! Running with any reasonable option like this and trying to prototype a syncing system is probably the right way to explore its feasibility in the details.

**Comment by pierreay** [comment]
Hello there. I'm adding my 50 cents here to a long-opened issue.

Practically speaking, I'm using the following setup since 1 year intensively without any issues:
* I'm syncing my Beets library (1 TB of music, mainly lossless) using [git-annex](https://git-annex.branchable.com/) across multiple computers and external hard drive. It allows me to grab my entire library or a part of it depending on the current device space, and even adding new files to it even if the library content is not present on my disk (see the unique features of this software).
* I'm syncing my Beets database (the SQLite database) using [Syncthing](https://syncthing.net/) across multiple computers and my smartphone. It allows me to manipulate and query the database in any places. 
* I'm syncting my Beets converted library (using the Beets' `convert` plugin) (100 GB of music) again using [Syncthing](https://syncthing.net/) across multiple computers and my smartphone. It allows me to always have my full music (using lossy compression) even on low storage space devices.
Thus, I'm concluding that to me, Beets does not need any syncing mechanism.

Philosophically speaking, UNIX's KISS would encourage Beets to not implement such a feature. I see a lot of software that tries to (re-)implement syncing mechanisms in a bad way. Indeed, this can quickly grow to a hard problem (connecting clients, centralized or decentralized, handling corruptions, handling conflicts, achieving privacy without leaking information, using encryption correctly, etc.). Moreover, there is today many FOSS software that is doing this well.
By letting the user choosing its syncing solution, it would give more free time and ressources to Beets developers for developing Beets unique features, while letting the freedom of choosing its own solution to the user. However, Beets could made some improvements to help this last point:
* Describing viable syncing solutions inside its documentation, maybe with examples.
* Implement conflict resolution (e.g., by implementing a command capable of comparing and merging two databases that have diverged).
